SwiftCompilerSources/Sources/Optimizer/Utilities/Devirtualization.swift

```diff
@@ -0,0 +1,236 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Devirtualizes all value-type deinitializers of a `destroy_value`.
+///
+/// This may be a no-op if the destroy doesn't call any deinitializers.
+/// Returns true if all deinitializers could be devirtualized.
+func devirtualizeDeinits(of destroy: DestroyValueInst, _ context: some MutatingContext) -> Bool {
+  return devirtualize(destroy: destroy, context)
+}
+
+/// Devirtualizes all value-type deinitializers of a `destroy_addr`.
+///
+/// This may be a no-op if the destroy doesn't call any deinitializers.
+/// Returns true if all deinitializers could be devirtualized.
+func devirtualizeDeinits(of destroy: DestroyAddrInst, _ context: some MutatingContext) -> Bool {
+  return devirtualize(destroy: destroy, context)
+}
+
+private func devirtualize(destroy: some DevirtualizableDestroy, _ context: some MutatingContext) -> Bool {
+  let type = destroy.type
+  guard type.isMoveOnly && type.selfOrAnyFieldHasValueDeinit(in: destroy.parentFunction) else {
+    return true
+  }
+  precondition(type.isNominal, "non-copyable non-nominal types not supported, yet")
+
+  let result: Bool
+  if type.nominal.hasValueDeinit && !destroy.hasDropDeinit {
+    guard let deinitFunc = context.lookupDeinit(ofNominal: type.nominal) else {
+      return false
+    }
+    destroy.createDeinitCall(to: deinitFunc, context)
+    result = true
+  } else {
+    // If there is no deinit to be called for the original type we have to recursively visit
+    // the struct fields or enum cases.
+    if type.isStruct {
+      result = destroy.devirtualizeStructFields(context)
+    } else {
+      precondition(type.isEnum, "unknown nominal value type")
+      result = destroy.devirtualizeEnumPayloads(context)
+    }
+  }
+  context.erase(instruction: destroy)
+  return result
+}
+
+// Used to dispatch devirtualization tasks to `destroy_value` and `destroy_addr`.
+private protocol DevirtualizableDestroy : UnaryInstruction {
+  func createDeinitCall(to deinitializer: Function, _ context: some MutatingContext)
+  func devirtualizeStructFields(_ context: some MutatingContext) -> Bool
+  func devirtualizeEnumPayload(enumCase: EnumCase, in block: BasicBlock, _ context: some MutatingContext) -> Bool
+  func createSwitchEnum(atEndOf block: BasicBlock, cases: [(Int, BasicBlock)], _ context: some MutatingContext)
+}
+
+private extension DevirtualizableDestroy {
+  var type: Type { operand.value.type }
+
+  var hasDropDeinit: Bool { operand.value.lookThoughOwnershipInstructions is DropDeinitInst }
```

Comment:
> hasDropDeinit should not be called on destroy_addr

Thanks! you are right.

> drop_deinit on an address does nothing but prevent the optimizer from reintroducing destroy_addr.

See my comments in https://github.com/apple/swift/pull/70052

> We should have a separate drop_deinit_addr since it has different semantics.

yes